File: LinkedList/python/problem_430_flatten_multilevel_doubly_linked_list.py
"""Problem 430: Flatten a Multilevel Doubly Linked List - Medium
Flatten multilevel doubly linked list with child pointers
Approach: DFS - process child before continuing
Time: O(n), Space: O(n) for recursion
"""
class Node:
    def __init__(self, val=0, prev=None, next=None, child=None):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child

class Solution:
    def flatten(self, head: Node) -> Node:
        if not head:
            return None
        self.flatten_dfs(head)
        return head
    
    def flatten_dfs(self, node: Node) -> Node:
        current = node
        last = None
        
        while current:
            next_node = current.next
            
            if current.child:
                child_last = self.flatten_dfs(current.child)
                
                # Connect current with child
                current.next = current.child
                current.child.prev = current
                
                # Connect child's tail with next
                if next_node:
                    child_last.next = next_node
                    next_node.prev = child_last
                
                current.child = None
                last = child_last
            else:
                last = current
            
            current = next_node
        
        return last

File: LinkedList/java/Problem430_FlattenMultilevelDoublyLinkedList.java
/**
 * Problem 430: Flatten a Multilevel Doubly Linked List
 * Difficulty: Medium
 * 
 * Flatten a multilevel doubly linked list where nodes may have a child pointer.
 * 
 * Approach: DFS - when encounter child, process it before continuing
 * Time: O(n), Space: O(n) for recursion
 */
class Node {
    public int val;
    public Node prev, next, child;
}

public class Problem430_FlattenMultilevelDoublyLinkedList {
    public Node flatten(Node head) {
        if (head == null) return null;
        flattenDFS(head);
        return head;
    }
    
    private Node flattenDFS(Node node) {
        Node current = node;
        Node last = null;
        
        while (current != null) {
            Node next = current.next;
            
            if (current.child != null) {
                Node childLast = flattenDFS(current.child);
                
                // Connect current with child
                current.next = current.child;
                current.child.prev = current;
                
                // Connect child's tail with next
                if (next != null) {
                    childLast.next = next;
                    next.prev = childLast;
                }
                
                current.child = null;
                last = childLast;
            } else {
                last = current;
            }
            
            current = next;
        }
        
        return last;
    }
}
